feat: support helm dryrun=server

holyspectral opened the pull request:
<!--

Welcome to the Operator SDK! Before contributing, make sure to:

- Read the contributing guidelines https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD
- Rebase your branch on the latest upstream master
- Link any relevant issues, PR's, or documentation
- Check that the commit message is concice and helpful:
    - When fixing an issue, add "Closes #<ISSUE_NUMBER>"
    - Sign your commit https://github.com/apps/dco
- Follow the below checklist if making a user-facing change 

Note, the location for ansible operator related logic has changed. For ansible operator related changes, please create the Pull Request in https://github.com/operator-framework/ansible-operator-plugins 

-->

**Description of the change:**

In this PR, new helm option, `--dry-run=server` is introduced, so `lookup()` can be used with helm operator. 

To enable this feature, specify this in watches.yaml
```
    dryrunOption: server
```



**Motivation for the change:**

`lookup()` in helm is a very useful command.  However, it's not supported by `helm xxx --dry-run` and `helm template` before 3.13.  This causes issues in some tools, e.g., argocd and helm operator.  This could cause repetitive reconcile or frequent certificate changes depending on how `lookup()` is used. 

Starting in Helm 3.13, helm provides a new option dryrun=server (https://github.com/helm/helm/pull/9426), which enabled users to contact API server when running helm.  This way, `lookup()` can get result from API server and work as it's intended to be. 



**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

everettraven commented on internal/helm/controller/controller.go:
```diff
@@ -52,6 +52,7 @@ type WatchOptions struct {
 	SuppressOverrideValues  bool
 	MaxConcurrentReconciles int
 	Selector                metav1.LabelSelector
+	DryrunOption            string
```
Nit: CamelCase:
```suggestion
	DryRunOption            string
```

everettraven commented on internal/helm/controller/controller.go:
```diff
@@ -66,6 +67,7 @@ func Add(mgr manager.Manager, options WatchOptions) error {
 		ReconcilePeriod:        options.ReconcilePeriod,
 		OverrideValues:         options.OverrideValues,
 		SuppressOverrideValues: options.SuppressOverrideValues,
+		DryrunOption:           options.DryrunOption,
```
Same nit as above

everettraven reviewed (commented):
I think this change looks reasonable. Would you be open to adding some test cases (either unit or e2e) that utilizes this new option to perform an action that would be allowed by this change?

holyspectral commented:
Fixed the CamelCase and added a simple test case.

everettraven reviewed (approved):
The changes look good to me but looks like there are some CI failures that need to be addressed. I think doing a `make generate` and committing the changes should fix the failure

holyspectral commented:
Hi @everettraven I rebased to master branch and I can reproduce the error on my local.  As your recommendation, I added these files in my latest commit and it seems to fix the issue.  However, I'm not entirely sure about changing `.cncf-maintainers`.
```
.cncf-maintainers
testdata/go/v4/memcached-operator/Makefile
testdata/go/v4/monitoring/memcached-operator/Makefile
testdata/helm/memcached-operator/Makefile
```

Could you help to double confirm if it's okay?  Thanks!

everettraven commented:
> Hi @everettraven I rebased to master branch and I can reproduce the error on my local. As your recommendation, I added these files in my latest commit and it seems to fix the issue. However, I'm not entirely sure about changing `.cncf-maintainers`.
> 
> ```
> .cncf-maintainers
> testdata/go/v4/memcached-operator/Makefile
> testdata/go/v4/monitoring/memcached-operator/Makefile
> testdata/helm/memcached-operator/Makefile
> ```
> 
> Could you help to double confirm if it's okay? Thanks!

@holyspectral Looks like that `.cncf-maintainers` change should have been made as part of https://github.com/operator-framework/operator-sdk/pull/6711 but was missed and generated as part of your PR. Should be OK

everettraven commented:
According to the DCO check one of the commits is missing a sign-off: https://github.com/operator-framework/operator-sdk/pull/6691/checks?check_run_id=23201517452

holyspectral commented:
Looks like there is still some error due to docs:
```
Checking 784 external links...
Ran on 171 files!


- /target/docs/best-practices/pod-security-standards/index.html
  *  External link https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml failed: 404 No error
- /target/docs/building-operators/golang/tutorial/index.html
  *  External link https://github.com/operator-framework/operator-sdk/blob/latest/testdata/go/v3/memcached-operator/controllers/memcached_controller.go failed: 404 No error
- /target/docs/upgrading-sdk-version/v1.18.0/index.html
  *  External link https://github.com/operator-framework/operator-sdk/commit/324ca13a994c7d7749708d550c0ea60d1df4cd0d failed: response code 0 means something's wrong.
             It's possible libcurl couldn't connect to the server or perhaps the request timed out.
             Sometimes, making too many requests at once also breaks things.
             Either way, the return message (if any) from the server is: Failed sending data to the peer
```

These are potentially related to https://github.com/operator-framework/operator-sdk/pull/6664 since testdata/go/v3 is removed.

holyspectral commented:
`v3/memcached-operator` are all changed to `v4/memcached-operator` now.  My internet connection is not good enough to complete `make test-docs` in my local, but at least no 404 anymore.

@everettraven can you help to take a look and rerun the test?  Thanks!

hkraal commented:
@everettraven It seems that changes from this PR aren't reflected at https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/ as [the memcached_controller link](https://github.com/operator-framework/operator-sdk/pull/6691/commits/3e23d36622dacc18d61d5a1c2d16da4d8560a1a7#diff-c6d77e2b8138b3ff32cb4bbacba02b794f9a602ce138373c764f07c80038a690R546) is still pointing to v3.

Is there any chance to update the docs?
